runtime/src/iree/hal/device.h

```diff
@@ -151,6 +151,87 @@ enum iree_hal_write_flag_bits_t {
+  // The application itself must ensure that any references captured by the call
+  // (user_data or args) are valid until the callback has completed.
+  //
+  // This is intended primarily for use as an optimization for custom signaling
+  // behavior or notifications.
+  IREE_HAL_HOST_CALL_FLAG_NON_BLOCKING = 1ull << 0,
+
+  // Hints that the host call is expected to be very short and that the issuing
+  // queue may want to spin (possibly with backoff) until the host call has
+  // signaled completion.
+  IREE_HAL_HOST_CALL_FLAG_WAIT_ACTIVE = 1ull << 1,
+
+  // Hints that the host call does not require the device to flush/invalidate
+  // caches. Use if the call does not consume any device resources that may have
+  // been produced but not yet flushed to host memory and does not produce any
+  // device resources that will be consumed without invalidation.
+  IREE_HAL_HOST_CALL_FLAG_RELAXED = 1ull << 2,
+};
+
+// Provides context to a host call about where it was made from as well as any
+// additional data requested.
+typedef struct iree_hal_host_call_context_t {
+  // The device the call was issued on.
+  iree_hal_device_t* device;
+  // The queue the call was issued on.
+  // This is guaranteed to be equal-to or a subset-of the queue affinity
+  // provided when the call was enqueued. Implementations are allowed to pick a
+  // single queue to call the operation on and block that or block entire groups
+  // of queues if there is some internal aliasing that introduces progress
+  // issues if only one queue is treated as blocked.
+  iree_hal_queue_affinity_t queue_affinity;
+  // A list of semaphores that must be signaled once the call has completed.
+  // Omitted if IREE_HAL_HOST_CALL_FLAG_NON_BLOCKING was requested.
+  //
+  // The list lives on the stack and must be copied and each semaphore retained
+  // if the call function does not immediately signal them inline. Asynchronous
+  // completion would clone the list, retain the semaphores, fire off the async
+  // operation, and then upon completion signal the semaphores and release them.
+  iree_hal_semaphore_list_t signal_semaphore_list;
+} iree_hal_host_call_context_t;
+
+// Executes a user-requested host call in queue order.
+// If the call succeeds and returns OK the semaphores will be signaled and
+// otherwise they will be failed. In non-blocking mode any error returned is
+// ignored and no semaphores are available.
+//
+// To implement asynchronous callbacks the signal_semaphore_list provided in
+// |context| should be cloned (list of pointers and retains on semaphores) and
+// stored for later signaling. The callback must return IREE_STATUS_DEFERRED to
+// indicate the asynchronous operation and when the operation has completed use
+// iree_hal_semaphore_list_signal or iree_hal_semaphore_list_fail based on
+// result.
+typedef iree_status_t(IREE_API_PTR* iree_hal_host_call_fn_t)(
+    void* user_data, const uint64_t args[4],
+    iree_hal_host_call_context_t* context);
```

Comment:
What is the 4 in uint64_t args[4]? I see that same number where iree_hal_host_call_t is used.

In the test I see values like `uint64_t args[4] = {10, 20, 30, UINT64_MAX};`. Is `4` just a fixed number of 64 bit arguments? Can a function take 0-4 arguments and ignore the values?

Reply:
Functions always take 4 args and must pass them (the `[4]` in the function argument declaration says that the function will read 4 elements) but may decide to use none of them if it wants.